src/oatpp/network/server/Server.cpp

```
@@ -30,47 +30,66 @@
-const v_int32 Server::STATUS_RUNNING = 1;
-const v_int32 Server::STATUS_STOPPING = 2;
-const v_int32 Server::STATUS_DONE = 3;
+const v_int32 Server::STATUS_STARTING = 1;
+const v_int32 Server::STATUS_RUNNING = 2;
+const v_int32 Server::STATUS_STOPPING = 3;
+const v_int32 Server::STATUS_DONE = 4;
 
 Server::Server(const std::shared_ptr<ServerConnectionProvider>& connectionProvider,
                const std::shared_ptr<ConnectionHandler>& connectionHandler)
   : m_status(STATUS_CREATED)
   , m_connectionProvider(connectionProvider)
   , m_connectionHandler(connectionHandler)
+  , m_blocking(false)
 {}
 
-void Server::mainLoop(){
+void Server::mainLoop(Server *instance){
   
-  setStatus(STATUS_CREATED, STATUS_RUNNING);
+  instance->setStatus(STATUS_STARTING, STATUS_RUNNING);
 
   std::shared_ptr<const std::unordered_map<oatpp::String, oatpp::String>> params;
 
-  while(getStatus() == STATUS_RUNNING) {
+  while(instance->getStatus() == STATUS_RUNNING) {
     
-    auto connection = m_connectionProvider->getConnection();
+    auto connection = instance->m_connectionProvider->getConnection();
 
     if (connection) {
-      if(getStatus() == STATUS_RUNNING){
-        m_connectionHandler->handleConnection(connection, params /* null params */);
+      if(instance->getStatus() == STATUS_RUNNING){
+        instance->m_connectionHandler->handleConnection(connection, params /* null params */);
       } else {
         OATPP_LOGD("Server", "Already stopped. Closing connection...");
       }
     }
     
   }
-  
-  setStatus(STATUS_DONE);
-  
+
+  instance->setStatus(STATUS_DONE);
+
 }
 
-void Server::run(){
-  mainLoop();
+void Server::run(bool blocking) {
+  if (blocking) {
+    m_blocking = blocking;
+    setStatus(STATUS_CREATED, STATUS_STARTING);
+    mainLoop(this);
+  } else {
+    switch (getStatus()) {
+      case STATUS_STARTING:
+      case STATUS_RUNNING:
+        // Maybe throw an exception here instead of just silently return?
```

Maybe throw an exception here instead of just silently return?